src/java/org/apache/cassandra/cql3/functions/types/VectorCodec.java

```diff
@@ -0,0 +1,211 @@
+
+        @Override
+        public boolean isValueLengthFixed()
+        {
+            return true;
+        }
+
+        @Override
+        public ByteBuffer serialize(List<E> value, ProtocolVersion protocolVersion) throws InvalidTypeException
+        {
+            if (value == null || type.getDimensions() <= 0)
+                return null;
+
+            ByteBuffer[] valueBuffs = new ByteBuffer[type.getDimensions()];
+            Iterator<E> values = value.iterator();
+            int allValueBuffsSize = 0;
+            for (int i = 0; i < type.getDimensions(); ++i)
+            {
+                ByteBuffer valueBuff = subtypeCodec.serialize(values.next(), protocolVersion);
+                allValueBuffsSize += valueBuff.limit();
+                valueBuff.rewind();
+                valueBuffs[i] = valueBuff;
+            }
+
+            // Since we already did an early return for <= 0 dimensions above
+            assert valueBuffs.length > 0;
+
+            ByteBuffer rv = ByteBuffer.allocate(allValueBuffsSize);
+            for (int i = 0; i < type.getDimensions(); ++i)
+            {
+                rv.put(valueBuffs[i]);
+            }
+            rv.flip();
+            return rv;
+        }
+
+        @Override
+        public List<E> deserialize(ByteBuffer bytes, ProtocolVersion protocolVersion) throws InvalidTypeException
+        {
+            if (bytes == null || bytes.remaining() == 0)
+                return null;
+
+            // Determine element size by dividing count of remaining bytes by number of elements.
+            // This should have a remainder of zero since all elements are of the same fixed size.
+            int elementSize = Math.floorDiv(bytes.remaining(), type.getDimensions());
+            assert bytes.remaining() % type.getDimensions() == 0
+            : String.format("Expected elements of uniform size, observed %d elements with total bytes %d",
+                            type.getDimensions(), bytes.remaining());
+
+            ImmutableList.Builder<E> values = ImmutableList.builder();
+            for (int i = 0; i < type.getDimensions(); ++i)
+            {
+                ByteBuffer slice = bytes.slice();
+                slice.limit(elementSize);
+                values.add(subtypeCodec.deserialize(slice, protocolVersion));
+                bytes.position(bytes.position() + elementSize);
+            }
+
+            // Restore the input ByteBuffer to its original state
+            bytes.rewind();
+
+            return values.build();
+        }
+    }
+
+    /**
+     * {@link VectorCodec} for vectors of elements using a varaible-length encoding.
+     */
+    private static class VariableLength<E> extends VectorCodec<E>
+    {
+        public VariableLength(VectorType type, TypeCodec<E> subtypeCodec)
+        {
+            super(type, subtypeCodec);
+        }
+
+        @Override
+        public ByteBuffer serialize(List<E> values, ProtocolVersion version) throws InvalidTypeException
```

Comment:
I was trying to think of a way to encode all sizes upfront, but only the deltas so save some bytes. But we don't support negative VInts...

Reply:
Any improvement on how vectors are encoded should be done in `VectorType`, the `VectorCodec` proposed here is just the driver codec mimicking it. That kind of improvement, if it's worth it, would also apply to collections.